### ISSUE TITLE
Multiple values for a metadata.

### ISSUE THREAD

**Issue by jahagirdar** [beancount/beancount#594]
In my workflow, All invoices are uploaded to google drive and a link to the invoice is added to the transaction via `url: "..."` metadata.
In some cases a single transaction can have multiple invoices, I tried capturing it 

1. via 2 url entries got error of `Duplicate posting metadata field: KeyValue(key='url', value='https://drive.google.com...`
2. Via a single url entry with 2 values and got `syntax error, unexpected COMMA,`

The metadata section of the document specifies strings (plural) for the valid value field, What is the right way to record the transaction?

**Comment by dnicolodi** [comment]
The current data Beancount data model does not allow for list valued metadata entries. Metadata is stored in a dictionary and thus metadata keys cannot be repeated. I think that the syntax and the data model can be extended to allow lists as metadata values, but this would have to be implemented in Beancount v3, which realistically will not be released soon. If you want to store multiple URLs under a single metadata key, you can put them in the same string separated by whitespace, ie:
```
2020-12-24 * "Foo"
    url: "https://foo.example https://bar.example"
    Assets:Cash 1.0
    Expenses:Foo
```
and unpack the string as a list of URLs when you retrieve the metadata value. URLs cannot contain spaces, thus this should work fine. Alternatively, you can associate metadata to the postings instead than to the transaction:
```
2020-12-24 * "Foo"
    Assets:Cash 1.0
        url: "https://foo.example"
    Expenses:Foo
        url: "https://bar.example"
```

**Comment by blais** [comment]
On Thu, Dec 24, 2020, 06:13 Daniele Nicolodi <notifications@github.com>
wrote:

> The current data Beancount data model does not allow for list valued
> metadata entries. Metadata is stored in a dictionary and thus metadata keys
> cannot be repeated. I think that the syntax and the data model can be
> extended to allow lists as metadata values, but this would have to be
> implemented in Beancount v3, which realistically will not be released soon.
>
+1
No big syntax changes in v2
Focus is all on v3

Even then it automatically creating lists on the presence of multiple
values would result in sometimes lists sometimes single objects, not super
consistent. Perhaps introducing a literal list syntax in branching would be
better.

Actually we could use that in other places, e.g. list of currencies in Open
directive.
Hmm


If you want to store multiple URLs under a single metadata key, you can put
> them in the same string separated by whitespace, ie:
>
> 2020-12-24 * "Foo"
>     url: "https://foo.example https://bar.example"
>     Assets:Cash 1.0
>     Expenses:Foo
>
> and unpack the string as a list of URLs when you retrieve the metadata
> value. URLs cannot contain spaces, thus this should work fine.
>
Nice workaround
+1




Alternativelyyyy, you can associate metadata to the postings instead than
> to the transaction:
>
2020-12-24 * "Foo"

**Comment by dnicolodi** [comment]
> Even then it automatically creating lists on the presence of multiple values would result in sometimes lists sometimes single objects, not super consistent.

I agree, this is not ideal and gets messy to handle quickly.  Although, there is no mechanism in place that asserts that metadata entries with the same key on different objects have the same type, thus technically, also now the code should be ready to handle metadata with different types.

> Perhaps introducing a literal list syntax in branching would be better.

I think that allowing a comma separated list of values is not too difficult. Allowing a trailing comma would give the possibility to specify a list of one element, allowing for consistent types independently of the number of entries in the list. One thing to decide is whether to allow lists of heterogeneous types (ie mixing decimal values, strings, etc...).

**Comment by blais** [comment]
> > Even then it automatically creating lists on the presence of multiple values would result in sometimes lists sometimes single objects, not super consistent.
> 
> I agree, this is not ideal and gets messy to handle quickly. Although, there is no mechanism in place that asserts that metadata entries with the same key on different objects have the same type, thus technically, also now the code should be ready to handle metadata with different types.

Good point indeed.

> 
> > Perhaps introducing a literal list syntax in branching would be better.
> 
> I think that allowing a comma separated list of values is not too difficult. Allowing a trailing comma would give the possibility to specify a list of one element, allowing for consistent types independently of the number of entries in the list. One thing to decide is whether to allow lists of heterogeneous types (ie mixing decimal values, strings, etc...).

I'm thinking about adding a more general construct at the parser level, for all lists of things, e.g.
```
[..., ..., ...]
```

This way you could
```
  2021-01-15 * "Return gifts"
    doc: ["receipt1", "receipt2", "receipt3"]
```
but also
```
  1980-01-01 open Assets:Cash     [USD, CAD]
```
WDYT?

**Comment by dnicolodi** [comment]
> > I agree, this is not ideal and gets messy to handle quickly. Although, there is no mechanism in place that asserts that metadata entries with the same key on different objects have the same type, thus technically, also now the code should be ready to handle metadata with different types.
> 
> Good point indeed.

It just came to my mind that it would not be difficult to have a plugin that enforces consistent types for metadata.

> I'm thinking about adding a more general construct at the parser level, for all lists of things

This is similar to what I had in mind, minus the square brackets. I haven't tried, but I suspect that having something like:

```
  2021-01-15 * "Return gifts"
    doc: "receipt1", "receipt2", "receipt3"
```
would be as easy to implement in the grammar. A list of one element would simply be:
```
  2021-01-15 * "Return gifts"
    doc: "receipt1",
```
similarly to how tuples are parsed in Python.

**Comment by yagebu** [comment]
I agree that a more general construct for lists could be useful and I'd be for the more explicit `[...]` syntax. I've always found the trailing comma in a single-entry Python tuple to be a fragile way to create a sequence structure. There's also other places beside the Beancount parser where this could be relevant (there's the grammar for the Fava editor for example and then also grammar-aware autocompletion in Fava in other places), so having a syntax where I can decide without lookahead whether something's a list sounds preferable to me.
